text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java

```diff
@@ -0,0 +1,276 @@
+            this.ansi.a(translatedContent);
+            this.ansi.a(Formats.SOFT_RESET);
+          }
+        }
+      }
+      final List<Component> children = component.children();
+      if(!children.isEmpty()){
+        final Style childrenStyle = new Style(style);
+        for(final Component child : children){
+          this.append(child, childrenStyle);
+          childrenStyle.set(style);
+        }
+      }
+    }
+
+    void append(final @NonNull TextFormat format) {
+      final Ansi a = ANSIComponentSerializerImpl.this.toANSI(format);
+      if(a != null){
+        this.ansi.a(a);
+      }
+    }
+
+    private final class Style {
+
+      private @Nullable TextColor color;
+      private final Set<TextDecoration> decorations;
+
+      Style(){
+        this.decorations = EnumSet.noneOf(TextDecoration.class);
+      }
+
+      Style(final @NonNull Style that){
+        this.color = that.color;
+        this.decorations = EnumSet.copyOf(that.decorations);
+      }
+
+      void set(final @NonNull Style that){
+        this.color = that.color;
+        this.decorations.clear();
+        this.decorations.addAll(that.decorations);
+      }
+
+      void apply(final @NonNull Component component){
+        final TextColor color = component.color();
+        if(color != null){
+          this.color = color;
+        }
+
+        for(final TextDecoration decoration : DECORATIONS){
+          switch(component.decoration(decoration)){
+            case TRUE:
+              this.decorations.add(decoration);
+              break;
+            case FALSE:
+              this.decorations.remove(decoration);
+              break;
+          }
+        }
+      }
+
+      void applyFormat() {
+        // If color changes, we need to do a full reset
+        if(this.color != Porridge.this.style.color){
+          this.applyFullFormat();
+          return;
+        }
+
+        // Does current have any decorations we don't have?
+        // Since there is no way to undo decorations, we need to reset these cases
+        if(!this.decorations.containsAll(Porridge.this.style.decorations)){
+          this.applyFullFormat();
+          return;
+        }
+
+        // Apply new decorations
+        for(final TextDecoration decoration : this.decorations){
+          if(Porridge.this.style.decorations.add(decoration)){
+            Porridge.this.append(decoration);
+          }
+        }
+      }
+
+      private void applyFullFormat() {
+        if(this.color != null){
+          Porridge.this.append(this.color);
+        }else{
+          Porridge.this.append(Reset.INSTANCE);
+        }
+        Porridge.this.style.color = this.color;
+
+        for(final TextDecoration decoration : this.decorations){
+          Porridge.this.append(decoration);
+        }
+
+        Porridge.this.style.decorations.clear();
+        Porridge.this.style.decorations.addAll(this.decorations);
+      }
+    }
+  }
+
+  enum Reset implements TextFormat {
+    INSTANCE
+  }
```

Comment:
Shouldn't this be in `Formats.java`?

Reply:
Its really only used privately in the StyleSetter.